feat: DP-128277 footer slot for DtDropdown

tiagox opened the pull request:
# DP-128277 footer slot for DtDropdown

## Obligatory GIF (super important!)

![Obligatory GIF](https://media2.giphy.com/media/v1.Y2lkPTc5MGI3NjExZmo5bjIyN3p1cjF5NXRpZjQwZTJ1ZXVmMzg4ZWZzNnRvY2pvNTJvayZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/fbL0RLU92W7Oo/giphy.gif)

## :hammer_and_wrench: Type Of Change

These types will increment the version number on release:

- [ ] Fix
- [x] Feature
- [ ] Performance Improvement
- [ ] Refactor

These types will not increment the version number, but will still deploy to documentation site on release:

- [ ] Documentation
- [ ] Build system
- [ ] CI
- [ ] Style (code style changes, not css changes)
- [ ] Test
- [ ] Other (chore)

## :book: Jira Ticket

DP-128277

## :book: Description

Add the ability to provide a footer section into a Dropdown. The `#footer` will be the footer of the underlying `DtPopover` in the `DtDropdown`.

## :pencil: Checklist

For all PRs:

- [ ] I have ensured no private Dialpad links or info are in the code or pull request description (Dialtone is a public repo!).
- [ ] I have reviewed my changes.
- [ ] I have added all relevant documentation.
- [ ] I have considered the performance impact of my change.

For all Vue changes:

- [ ] I have added / updated unit tests.
- [ ] I have made my changes in Vue 2 and Vue 3. Note: you may sync your changes from Vue 2 to Vue 3 (or vice versa) using the `./scripts/dialtone-vue-sync.sh` script. Read docs here: [Dialtone Vue Sync Script](../packages/dialtone-vue3/.github/CONTRIBUTING.md#dialtone-vue-sync-script)
- [ ] I have validated components with a screen reader.
- [ ] I have validated components keyboard navigation.

For all CSS changes:

- [ ] I have used design tokens whenever possible.
- [ ] I have considered how this change will behave on different screen sizes.
- [ ] I have visually validated my change in light and dark mode.
- [ ] I have used gap or flexbox properties for layout instead of margin whenever possible.

## :camera: Screenshots / GIFs

![image](https://github.com/user-attachments/assets/85a80efc-0cfc-436f-bd1f-150bf2e344f5)

## :link: Sources

Figma link for the feature: https://www.figma.com/design/dkjx25f3qXIJeTYbDBq6O3/%5BDP-116475%5D-Agent-log-in-%2F-log-out?node-id=5408-48888&m=dev

francisrupert commented on lib/build/less/components/modal.less:
```diff
@@ -139,18 +137,18 @@
   margin: 0 !important;
-  padding: var(--su24) var(--su24) 0;
+  padding: var(--su24) var(--su24) 0 !important;
   color: var(--modal-header--fc);
   font-weight: var(--fw-bold);
-  font-size: var(--fs24);
+  font-size: var(--fs24) !important;
```
Is this a bit too brute-force? I thought the real culprit was VuePress' over-specific CSS selectors leaking into our components? 

Can VuePress' CSS be updated rather than updating the DT component itself?

The age-old "avoid `!important`" is the go-to thinking here, particularly since we're deliberate about `!important` for Utilities.

josedialpad replied:
Yeah I agree and the first approach I've tried was to adapt the styles not to override the modal's style, and I found that we're overriding the styles here https://github.com/dialpad/dialtone/blob/staging/docs/.vuepress/styles/palette.scss#L11-L24 that's in our side and can be modified, but also the Vuepress default theme (which we're using) has a set of styles we cannot modify since they're coming from the `@vuepress` package https://github.com/vuepress/vuepress-next/blob/main/packages/%40vuepress/theme-default/src/client/styles/layout.scss#L123-L152

What I found in Vuepress v1 is the `eject` command to copy the default theme source into the project so that way we can customize it but that implies the theme will stop to receive any update and will require manual update. Also the `eject` command was removed in the v2 https://v2.vuepress.vuejs.org/guide/migration.html#cli-change:~:text=%23-,eject%20command,-Removed.

braddialpad replied:
Is there a way we can change the specificity of the CSS in this component to make this work without important? 

Additionally, looks like `.theme-default-content` is the problematic class, is there a way we can exclude `.theme-default-content` from affecting our examples at the Vuepress level? Is it possible to use `:not()` or something like that?

josedialpad replied:
> Additionally, looks like .theme-default-content is the problematic class, is there a way we can exclude .theme-default-content from affecting our examples at the Vuepress level? Is it possible to use :not() or something like that?

I couldn't find an easy way, to do it it requires changing the layout or extending the default theme.

> Is there a way we can change the specificity of the CSS in this component to make this work without important?

Updated to fix with specificity https://github.com/dialpad/dialtone/pull/635/commits/9aed7077a8e14cd4e973b7a288b32e4211423dc1

Basically, we need to override the `.d-modal__header` class just giving higher specificity to win the battle against the styles coming from the default theme that's not possible to modify https://github.com/vuepress/vuepress-next/blob/main/packages/%40vuepress/theme-default/src/client/styles/layout.scss#L123-L152

francisrupert replied:
Quite a workaround, nicely done. While it's a blatant code smell...

![image](https://user-images.githubusercontent.com/1165933/183106430-2241c791-dab2-49c3-8d08-822fc0af934d.png)

...it's as reasonable a workaround as any, better than `!important`. Hopefully we won't have to do this too often going forward, let alone if we haven't caught other instances of VuePress styles leaking in.

Very unfortunate VuePress didn't realize how much their style approach _will_ leak in. A significant redesign/rearchitecture of doc site is on the backburner, but nowhere near a priority. 

I approve this change, what say you @braddialpad?

braddialpad replied:
Yep I'm okay with this.

tiagox committed:
feat: DP-128277 add footer slot to dt-dropdown vue2

tiagox committed:
feat: DP-128277 add footer slot to dt-dropdown vue3

ninamarina committed:
fix to display the footer in vue 2

juliodialpad commented on packages/dialtone-vue2/components/dropdown/dropdown_variants.story.vue:
```diff
@@ -162,6 +162,60 @@
+            </dt-button>
+          </div>
+        </template>
+      </dt-dropdown>
```
A little clean up suggested, we don't really need all of those variables as the `variants` stories are meant to be `static` examples, we only keep the `:open="$attrs.open"` as it is needed for visual testing.

```suggestion
      <dt-dropdown
        class="d-mr8"
        :open="$attrs.open"
      >
        <template
          slot="anchor"
          slot-scope="{ attrs }"
        >
          <dt-button
            v-bind="attrs"
          >
            with footer
          </dt-button>
        </template>
        <template
          slot="list"
          slot-scope="{ close }"
        >
          <dt-list-item
            v-for="(item) in items"
            :key="item.id"
            role="menuitem"
            :navigation-type="$attrs.navigationType"
            @click="close"
          >
            {{ item.name }}
          </dt-list-item>
        </template>
        <template
          slot="footer"
          slot-scope="{ close }"
        >
          <div class="d-px12">
            <dt-button @click="close">
              Dropdown footer
            </dt-button>
          </div>
        </template>
      </dt-dropdown>
```

juliodialpad commented on packages/dialtone-vue3/components/dropdown/dropdown_variants.story.vue:
```diff
@@ -156,6 +156,40 @@
+        :placement="variant"
+        :padding="$attrs.padding"
+        :navigation-type="$attrs.navigationType"
+        :transition="$attrs.transition"
```
```suggestion
```

juliodialpad commented on packages/dialtone-vue3/components/dropdown/dropdown_variants.story.vue:
```diff
@@ -156,6 +156,40 @@
+        :navigation-type="$attrs.navigationType"
+        :transition="$attrs.transition"
+        :open="$attrs.open"
+        @highlight="$attrs.onHighlight"
```
```suggestion
```

juliodialpad commented on packages/dialtone-vue3/components/dropdown/dropdown_variants.story.vue:
```diff
@@ -156,6 +156,40 @@
+          <dt-button>
+            with footer
+          </dt-button>
+        </template>
```
Keep Vue 2 and Vue 3 code as close as possible

```suggestion
        <template #anchor="{ attrs }">
          <dt-button v-bind="attrs">
            with footer
          </dt-button>
        </template>
```

juliodialpad reviewed (changes requested):
Thanks for contributing! this is a really nice first time contribution, just a small changes requested.

1. Please make sure the slot works on the default story, add it to `dropdown_default.story.vue` and set it on `dropdown.stories.js` too please. You can refer to `DtPopover` component stories or send me a message directly if needed.

 Thanks!

tiagox committed:
Apply suggestions from code review

Co-authored-by: Julio Ortega <julio.ortega@dialpad.com>

tiagox committed:
add footer slot config for stories

tiagox committed:
add footer slot in default story

tiagox committed:
replace deprecated slot prop with shorthand syntax

Additionally, bring Vue2 and Vue3 versions closer

tiagox commented:
> Keep Vue 2 and Vue 3 code as close as possible

With this in mind I've replaced the deprecated `slot=""` as a prop for the shorthand syntax in Vue2.

Also, I believe I have addressed your comments:

- Footer added to the default story.
- `footer` slot added in prop definition for live editing in Storybook.

braddialpad commented on packages/dialtone-vue2/components/popover/popover.vue:
```diff
@@ -595,6 +595,10 @@ export default {
+
+    hasFooter () {
+      return this.$slots.footerContent || (this.$scopedSlots.footerContent && this.$scopedSlots.footerContent());
+    },
```
Hmm I don't believe this will be reactive because it's referencing $slots. Instead of a computed make this a method and call it from the template.

tiagox replied:
I've moved this to the methods section and it's not being reactive either. If I set the `footer` in storybook, then I need to make the component to re-create (going to another page and back to the dropdown) to see the new footer finally rendered.

I'm not sure why setting the slot `#footerContent` is not showing it in `$slots.footerContent`.

braddialpad replied:
hmmm just a sec going to take a look myself.

tiagox replied:
Here is the interaction

[scoped-slot-re-rendering.webm](https://github.com/user-attachments/assets/ce89cbc8-6424-4710-b415-945024f9db9c)

only applicable for Vue2. Vue3 works fine.

braddialpad replied:
It doesn't even work on the popover component on the currently live storybook, so I think that's where the problem is https://dialtone.dialpad.com/vue/index.html?path=/story/components-popover--default

braddialpad replied:
Okay so I've confirmed this issue, it is definitely a bit strange... If you don't initialize the slots with a value they just never update when entered in the UI, if you initialize them with anything, even an empty string, they will then update reactively, so I added this to "popover.stories.js"

```
args: {
    headerContent: '',
    footerContent: '',
    content: '',
  },
```

It's weird because props do not work this same way.

There is probably a better, more global fix we can do for this but it's really nothing to do with this PR. I pushed a fix for this case, you should be good to go.

Thanks for struggling through that with us.

braddialpad commented on packages/dialtone-vue2/components/dropdown/dropdown_default.story.vue:
```diff
@@ -52,6 +46,13 @@
+          Dropdown footer
+        </dt-button>
+      </div>
+    </template>
```
For this one you can just go

```
<template #footer>
  <div v-if="$attrs.footer" v-html="$attrs.footer" />
</template>
```

This allows the user to pass in whatever html tags they want to it from storybook (empty by default).

braddialpad commented on packages/dialtone-vue2/components/popover/popover.vue:
Do we also need to do this in the vue 3 popover?

tiagox replied:
I'm not sure. This was introduced because in Vue2, the `#footer` slot from the Dropdown wasn't working at all.

@ninamarina can you elaborate on this?

ninamarina replied:
Exactly, the way we check for slot content in Vue 3 is different, so the problem was only in Vue 2.

braddialpad replied:
sounds good suspected that might be the case. Just wanted to make sure we didn't miss copying something to vue 3.

braddialpad reviewed (changes requested):
Thanks for the PR!

tiagox committed:
convert hasFooter to a method

tiagox committed:
make footer slot configurable in Dropdown default story

braddialpad committed:
fix slots not rendering in popover

braddialpad committed:
fix on dropdown also

braddialpad reviewed (approved):
Looks good thanks!
